designs/flow-linter/README.md

```diff
@@ -0,0 +1,147 @@
+---
+state: draft
+---
+
+# Flow Linter
+
+## Summary
+Node-RED makes it easy to program by wiring nodes together.  But sometimes, because of its less restrictive nature, programmer may write flows that are hard to understand.  For example: 
+- Function node that has no name or description.  The reader need to read internal JavaScript code to understand what the node do,
+- Excessive amount of nodes in a single flow.
+
+And, some subtle differences in programming style cause readability low.
+- Size of flow on Flow Editor
+- Grid size
+- JavaScript coding style in function nodes.
+
+The flow linter provides a framework that automatically checks whether a flow is compliant with rules and conventions. 
+
+## Authors
+ - [@k-toumura](https://github.com/k-toumura)
+
+## Details
+
+### Use cases
+- To use the consistent coding rules in a team, the tool checks a flow before submit to a repository.
+- To prevent bugs, the tool checks a flow when the programmer requests it, or continuously checks as a background task of Flow Editor. 
+  - Editor displays suggestions to user: 'This HTTP-in node doesn't have corresponding HTTP-out node.', 'These nodes in this flow must preserve message properties set by HTTP-in', etc.
+- To prevent move/edit/delete node accidentally, some node position or property change can be restricted by rules.
+
+### Requirements / ideas
+- There are various rules/conventions, and each organization/community/etc. has different policies on it.  Because of this, the rules has to be pluggable and customizable.
+- Linter is used in both batch-style (e.g. command-line interface) and on-the-fly-style (e.g. integrated in Editor).  The core service of linter can be used in both usages.
+- Rule configuration can be exported as JSON format and can be included in flow.json so that developers can distribute flow templates with their own restrictions.
+- Use compatible format with other linter/tester tools for result output.
+
+### User interface
+
+#### Command-line interface (batch-style)
+![External CLI](external-cli.png)
+
+#### Integrated in Editor (in Sidebar, batch-style)
```

Comment:
We'll need more detailed design work around how the linter will be exposed in the editor - but that comes after we have the core design of the linter settled, so not (I believe) in scope for this initial design PR.

Reply:
For this initial PR, I'll add some additional descriptions for each UI patterns, and comments for necessity of further consideration.